### PR TITLE
Add parallelism within test files

### PR DESCRIPTION
**Pull request by zatkins-dev**
Adds parallelism within test cases as implemented in Ratel: https://gitlab.com/micromorph/ratel/-/merge_requests/629

See https://gitlab.com/micromorph/ratel/-/issues/134 for more discussion.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -38,21 +39,21 @@ noether-cpu:
...
     - echo "-------------- MFEM ----------------" && make -C $MFEM_DIR info
-    - make -k -j$NPROC_CPU BACKENDS="$BACKENDS_CPU" JUNIT_BATCH="cpu" junit search=mfem
+    - make -k -j$((NPROC_CPU / NPROC_POOL)) BACKENDS="$BACKENDS_CPU" JUNIT_BATCH="cpu" junit search=mfem
 # -- Nek5000 v19.0
     - export COVERAGE=0
     - cd .. && export NEK5K_VERSION=Nek5000-19.0 && { [[ -d $NEK5K_VERSION ]] || { git clone --depth 1 --branch v19.0 https://github.com/Nek5000/Nek5000.git $NEK5K_VERSION && cd $NEK5K_VERSION/tools && ./maketools genbox genmap reatore2 && cd ../..; }; } && export NEK5K_DIR=$PWD/$NEK5K_VERSION && export PATH=$NEK5K_DIR/bin:$PATH MPI=0 && cd libCEED
     - echo "-------------- Nek5000 -------------" && git -C $NEK5K_DIR describe --tags
-    - make -k -j$NPROC_CPU BACKENDS="$BACKENDS_CPU" JUNIT_BATCH="cpu" junit search=nek NEK5K_DIR=$NEK5K_DIR
+    - make -k -j$((NPROC_CPU / NPROC_POOL)) BACKENDS="$BACKENDS_CPU" JUNIT_BATCH="cpu" junit search=nek NEK5K_DIR=$NEK5K_DIR
```
Comment (jeremylt) [inline .gitlab-ci.yml:56]:
```suggestion
    - make -k -j$NPROC_CPU BACKENDS="$BACKENDS_CPU" JUNIT_BATCH="cpu" junit search=nek NEK5K_DIR=$NEK5K_DIR
```

Nek5K's build/run/ect script is painful.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -107,21 +109,21 @@ noether-rocm:
...
     - echo "-------------- MFEM ----------------" && make -C $MFEM_DIR info
-    - make -k -j$NPROC_GPU BACKENDS="$BACKENDS_GPU" JUNIT_BATCH="hip" junit search=mfem
+    - make -k -j$((NPROC_GPU / NPROC_POOL)) BACKENDS="$BACKENDS_GPU" JUNIT_BATCH="hip" junit search=mfem
 # -- Nek5000 v19.0
     - export COVERAGE=0
     - cd .. && export NEK5K_VERSION=Nek5000-19.0 && { [[ -d $NEK5K_VERSION ]] || { git clone --depth 1 --branch v19.0 https://github.com/Nek5000/Nek5000.git $NEK5K_VERSION && cd $NEK5K_VERSION/tools && ./maketools genbox genmap reatore2 && cd ../..; }; } && export NEK5K_DIR=$PWD/$NEK5K_VERSION && export PATH=$NEK5K_DIR/bin:$PATH MPI=0 && cd libCEED
     - echo "-------------- Nek5000 -------------" && git -C $NEK5K_DIR describe --tags
-    - make -k -j$NPROC_GPU BACKENDS="$BACKENDS_GPU" JUNIT_BATCH="hip" junit search=nek NEK5K_DIR=$NEK5K_DIR
+    - make -k -j$((NPROC_GPU / NPROC_POOL)) BACKENDS="$BACKENDS_GPU" JUNIT_BATCH="hip" junit search=nek NEK5K_DIR=$NEK5K_DIR
```
Comment (jeremylt) [inline .gitlab-ci.yml:126]:
```suggestion
    - make -k -j$NPROC_GPU BACKENDS="$BACKENDS_GPU" JUNIT_BATCH="hip" junit search=nek NEK5K_DIR=$NEK5K_DIR
```
same here and below

File: .gitlab-ci.yml (hunk 1 above)
Comment (zatkins-dev) [reply .gitlab-ci.yml]: Ah yes, I just dug into that and died a little

Comment (jeremylt) [reply .gitlab-ci.yml]: Your fortitude is stronger than mine. There was much weeping, gnashing of teeth, and dying inside when I was trying to set up Nek5K CI initially.

**Comment by jeremylt** [comment]
37 minutes 58 seconds (old) vs 17 minutes 14 seconds (new) for the slowest job. Seems like a win.

**Comment by zatkins-dev** [comment]
This seems really slow:
image

**Comment by jeremylt** [comment]
Yeah, its probably worth investigating if we can reproduce manually

**Comment by zatkins-dev** [comment]
> Yeah, its probably worth investigating if we can reproduce manually

I'm able to reproduce with `/gpu/magma/*` backends locally (which we aren't currently testing, btw)

**Comment by jeremylt** [comment]
Cool - I'd create an issue for follow-up since its tangential to this PR

**Comment by sebastiangrimberg** [comment]
> I'm able to reproduce with `/gpu/cuda/magma/*` backends locally (which we aren't currently testing, btw)

I'm not certain, but I imagine this is related to the very large RTC overhead that the `magma` backends have currently (see https://github.com/CEED/libCEED/issues/1252). Should be fixed by lazy compiation in https://github.com/CEED/libCEED/pull/1301.

**Comment by jeremylt** [comment]
Ahh, that could be the culprit. Stripping that out into a stand-alone PR we can merge sooner would probably help, if that's not an unreasonable amount of work.

**Review by jeremylt — APPROVED** [review]

**Comment by sebastiangrimberg** [comment]
> t. Stripping that out into a stand-alone PR we can merge sooner would probably help, if that's not an unreasonable amount of work.

I can definitely do this but (fingers crossed) we also seem very close in closing the remaining issue of performance degradation on AMD. So I would prefer to wait at least a few days to hear back and see if we can fix the issue and merge the whole thing at once rather than work on splitting everything out into a separate PR, if that sounds OK.

**Comment by jeremylt** [comment]
Sure, that works. One of these decades we'll all get good at making smaller PRs from the get-go :upside_down_face:

**Pull request merged**